fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java

```diff
@@ -898,69 +875,170 @@ private static OptExpression generateDecodeOExpr(DecodeContext context, List<Opt
+        public Void visitCall(CallOperator call, CouldApplyDictOptimizeContext context) {
             if (!call.getFunction().isCouldApplyDictOptimize()) {
-                return false;
+                context.hasUnsupportedOperator = true;
+                return null;
             }
-            return call.getChildren().stream().allMatch(scalarOperator -> scalarOperator.accept(this, null));
+
+            couldApply(call, context);
+            context.worthApplied |= context.couldAppliedOperator;
+            return null;
         }
 
         @Override
-        public Boolean visitBinaryPredicate(BinaryPredicateOperator predicate, Void context) {
-            if (predicate.getBinaryType() == EQ_FOR_NULL) {
-                return false;
-            }
-            if (!predicate.getChild(1).isConstant()) {
-                return false;
+        public Void visitBinaryPredicate(BinaryPredicateOperator predicate, CouldApplyDictOptimizeContext context) {
+            if (predicate.getBinaryType() == EQ_FOR_NULL || !predicate.getChild(1).isConstant() ||
+                    !predicate.getChild(0).isColumnRef()) {
+                context.couldAppliedOperator = false;
+                context.hasUnsupportedOperator = true;
+                return null;
             }
-            return predicate.getChild(0).isColumnRef();
+
+            predicate.getChild(0).accept(this, context);
+            context.worthApplied |= context.couldAppliedOperator;
+            return null;
         }
 
         @Override
-        public Boolean visitInPredicate(InPredicateOperator predicate, Void context) {
-            return predicate.getChild(0).isColumnRef() &&
-                    predicate.allValuesMatch(ScalarOperator::isConstantRef);
+        public Void visitInPredicate(InPredicateOperator predicate, CouldApplyDictOptimizeContext context) {
+            if (!predicate.allValuesMatch(ScalarOperator::isConstantRef) || !predicate.getChild(0).isColumnRef()) {
+                context.couldAppliedOperator = false;
+                context.hasUnsupportedOperator = true;
+                return null;
+            }
+
+            predicate.getChild(0).accept(this, context);
+            context.worthApplied |= context.couldAppliedOperator;
+            return null;
         }
 
         @Override
-        public Boolean visitIsNullPredicate(IsNullPredicateOperator predicate, Void context) {
-            return predicate.getChild(0).isColumnRef();
+        public Void visitIsNullPredicate(IsNullPredicateOperator predicate, CouldApplyDictOptimizeContext context) {
+            if (!predicate.getChild(0).isColumnRef()) {
+                context.couldAppliedOperator = false;
+                context.hasUnsupportedOperator = true;
+                return null;
+            }
+
+            predicate.getChild(0).accept(this, context);
+            context.worthApplied |= context.couldAppliedOperator;
```

Comment:
In fact, worthApplied is not the cumulative result of couldAppliedOperator.
eg: concat(dict1, dict2)
when we visit dict1 and dict2, the value of couldAppliedOperator is true.
The final result of worthApplied should be false. Because it is not worth using low-cardinality optimization in this case